Javadoc Enhancements (#28712)
gradle/gradle: JavadocStyleGuide.md

# Gradle Javadoc Style Guide

## 1.1 Formatting

### 1.1.1 General form

The basic formatting of Javadoc blocks is as seen in this example:

```java
/**

>  * Returns an Image object that can then be painted on the screen.
>  * The url argument must specify an absolute <a href="#{@link}">{@link URL}</a>. The name

Reviewer: ```suggestion
 * Returns an Image object that can then be painted on the screen.
 * <p>
 * The url argument must specify an absolute <a href="#{@link}">{@link URL}</a>. The name
```

I think we should always have one short sentence describing the "Thing", and then a newline. In this example, that makes total sense as well.

Reviewer: IDEA seems to disagree: 
<img width="683" alt="image" src="https://github.com/gradle/gradle/assets/423186/ecf6bef7-c7c0-4988-855a-c24889028c80">

Aka it doesn't see that there is a paragraph after the summary.

Reviewer 2: I thought no paragraph symbol was needed as well - and perhaps it isn't for compiled HTML javadoc - but if we can't get IDEA to figure out how to add paragraphs when Javadoc is written with only blank lines, we should require the `<p>`.  Viewing in the IDE is an important use case.

Author: Style guide has been updated, IntelliJ IDEA sees <p>, <br>, and * blank line as the same thing.

 * argument is a specifier that is relative to the url argument.
 * <p>
 * This method always returns immediately, whether or not the
 * image exists. When this applet attempts to draw the image on
 * the screen, the data will be loaded. The graphics primitives
 * that draw the image will incrementally paint on the screen.
 *
 * @param  url  an absolute URL giving the base location of the image
 * @param  name the location of the image, relative to the url argument
 * @return      the image at the specified URL

>  * @see         Image

Reviewer 3: All of these examples have the description text left-aligned. This can be very tedious when adding new parameters or changing the names of existing parameters. 

Consider the use-case of changing `name` to `imageName` (for whatever reason). Even though I just wanted to modify a single line, I would also need to modify three other lines. 

There are a few downsides to this:
1. Increases complexity of code review. Only `name` -> `imageName` changed, but three other unrelated lines show up in the code diff. 
2. Reduces accuracy of git history. Git history is important for determining when features got added, when bugs were introduced, or determining context of existing code. Causing unrelated code changes like this (renaming `name` affecting surrounding lines) can make it more difficult to track changes in git history, as initially it appears that `@param url` and `@return` changed when in reality their whitespace was modified due to the renaming of `name`. 

I have not seen any official notes in this page saying we _should_ have these text descriptions left-aligned, but all examples show this as the case. I do not think we should enforce alignment of these text blocks.

Reviewer 2: Please let's not do this.  Single spaces between `@param`, the name, and the desc.  Likewise for the other tags.

Author: <p>, <br> and and empty * line are all line break in IntelliJ IDEA. Style guide has been updated.

Author: Style guide has been updated to continue with the curent convention (no left alignment).

 */
public Image getImage(URL url, String name) {
   try {
       return getImage(new URL(url, name));
   } catch (MalformedURLException e) {
       return null;
   }
}
```

Basic formatting rules:

- The first line contains the begin-comment delimiter ( `/**`).
- Notice the inline tag `{@link URL}`, which converts to an HTML hyperlink pointing to the documentation for the URL class.

> - If you have more than one paragraph in the doc comment, separate the paragraphs with a `<p>` paragraph tag, as shown.

Reviewer 3: When I type `<p>` in a javadoc segment, IntelliJ automatically adds the corresponding `</p>` for me. For this reason, I have been surrounding each paragraph with `<p>text</p>`. 

I understand the official Oracle javadoc style guide says to use a single `<p>` between paragraphs, so if we choose to enforce this bullet, we should try to fix IntellIJ from adding `</p>` automatically. Otherwise, the IDE suggestions directly conflict with this style rule.

Reviewer 2: It doesn't do this for me...don't know how I configured this though.

Author: Settings -> Editor -> General -> Smart Keys -> "Insert closing tag on tag completion"

- Block tags must be added in order.
- The last line contains the end-comment delimiter ( `*/`).

So lines won't wrap, limit any doc-comment lines to 80 characters.

### 1.1.2 HTML

Javadoc permits only a subset of HTML tags: `"a", "abbr", "acronym", "address", "area", "b", "bdo", "big", "blockquote", "br", "caption", "cite", "code", "colgroup", "dd", "del", "dfn", "div", "dl", "dt", "em", "fieldset", "font", "h1", "h2", "h3", "h4", "h5", "h6", "hr", "i", "img", "ins", "kbd", "li", "ol", "p", "pre", "q", "samp", "small", "span", "strong", "sub", "sup", "table", "tbody", "td", "tfoot", "th", "thead", "tr", "tt", "u", "ul", "var"`.

Attempt to limit your usage of HTML to: `"h1", "h2", "h3", "h4", "em", "li", "ol", "p", "ul"`.

Some HTML element types allow authors to omit end tags: `<p>` and `<li>`.

### 1.1.3 Paragraphs

One blank line—that is, a line containing only the aligned leading asterisk (`*`)—appears between the last paragraph and before the group of block tags if present:

```java
/**
 * that draw the image will incrementally paint on the screen.
 *
 * @param  url  an absolute URL giving the base location of the image
 */
```

> Each paragraph except the first has `<p>` immediately before the first word, with no space after it:

Reviewer 2: 💭 I think we should require putting the `<p>`s on their own line to make them more visually obvious when editing the javadoc text.

```
/**
 * argument is a specifier that is relative to the url argument.
 * <p>
 * This method always returns immediately, whether or not the
 */

Author: Agreed. Style guide has been updated.

```java
/**
 * argument is a specifier that is relative to the url argument.
 * <p>This method always returns immediately, whether or not the
 */
```

Alternatively, a `<p>` can be placed on a separate line:

```java
/**
 * argument is a specifier that is relative to the url argument.
 * <p>
 * This method always returns immediately, whether or not the
 */
```

The inline tag `{@link URL}`, which converts to an HTML hyperlink pointing to the documentation for the URL class. 
This inline tag can be used anywhere that a comment can be written, such as in the text following block tags:

```java
/**

>  * The url argument must specify an absolute <a href="#{@link}">{@link URL}</a>. The name

Reviewer 3: This document references the `<a href="#{@link}">{@link URL}</a>` syntax a lot, but I hardly see it in practice in the real javadoc in `gradle/gradle`. 

Generally I much more often see `{@link ClassName}` without any `a href` tag. IMO we should focus on that format

Author: Styling guide has been updated.

 */
```

HTML tags for other block-level elements, such as `<ul>` or `<table>`, are not preceded with `<p>`.

### 1.1.4 Symbols

Entities for the less than symbol (<) and the greater than symbol (>) should be written as `&lt`; and `&gt;`. 
Similarly, the ampersand (&) should be written as `&amp;`.

```java
/**
 * The url argument must specify an absolute <a href="#{@link}">{@link URL}</a>. The name
 */
```
### 1.1.5 Links

The `{@link <class or method reference>}` tag is specifically used to link to the Javadoc of other classes and methods. 
This is an inline tag that converts to an HTML hyperlink pointing to the documentation of the given class or method reference

```java
/**

>  * This &amp; that.

Reviewer: I think this example got mixed up with the one from above.

Author: Fixed. Thank you.

 */
```

### 1.1.6 Code keywords, names, and variables

> Use `<code>...</code>` style for keywords and names including:

Reviewer 3: I much prefer `{@code ... }` for two main reasons:

1. It is more concise.
2. HTML characters are automatically-escaped. 

For example, I can write `{@code List<String>}` instead of `<code>List&lt;String&gt;</code>`

Reviewer 4: I think regardless of our decision both options should be mentioned in the style guide. For instance, we should explicitly say that using `{@code ...}` is preferred over the HTML equivalent or that it does not matter which one is used.

Author: Both are ok. Updated the style guide accordingly.

- Java keywords
- package names
- class names
- method names
- interface names
- field names
- argument names
- code examples (see section 1.3 below)

```java
/**
 * Use the <code>Project</code> instance to configure the project.
 */
```

### 1.1.7 Block tags

> Any of the standard "block tags" that are used appear in the order `@param`, `@return`, `@throw`s`, `@deprecated`, and these four types never appear with an empty description:

Reviewer 3: ```suggestion
Any of the standard "block tags" that are used appear in the order `@param`, `@return`, `@throws`, `@deprecated`, and these four types never appear with an empty description:
```

Author: Fixed. Thank you.

```java
/**
 * @param  url  an absolute URL giving the base location of the image
 * @param  name the location of the image, relative to the url argument
 * @return      the image at the specified URL
 * @see         Image
 */
```

Full list of tags (in order):

| # | Tag           | Usage                                                                          | Notes                                                                                      |
|---|---------------|--------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------|
| 1 | `@param`      | Methods and constructors only                                                  | Include this if applicable                                                                 |
| 2 | `@return`     | Methods only                                                                   | Include this if applicable                                                                 |

> | 3 | `@exception`  | Same as `@throws`                                                              | Include this if applicable                                                                 |

Reviewer 3: Can we put `@throws` in the first column and `Same as @exception` in the second? I see `@throws` MUCH more often then `@exception` and I think that should be the "main" tag to use here

Author: Sounds good. Done.

| 4 | `@see`        | Adds a “See Also” heading with a link or text entry that points to a reference | `@see string`<br>`@see <a href=”URL#value”>label</a>`<br>`@see package.class#member label` |
| 5 | `@since`      | Adds a “Since” heading                                                         | Include the Gradle version if applicable                                                   |
| 6 | `@deprecated` | Adds a comment indicating that this API should no longer be used               | Make sure to have an alternative API linked                                                |

Custom Gradle tags:

| # | Tag            | Usage                                        | Notes                                            |
|---|----------------|----------------------------------------------|--------------------------------------------------|

> | 1 | `@apiNote`     | Adds a “API Note” heading                    | Describe interesting details about the API       |

Reviewer 2: ❗ This is missing `@apiSpec`.

Author: Added. Thank you.

> | 2 | `@implSpec`    | Adds a “Implementation Requirements” heading | Describe what we expect the plugin author to do  |

Reviewer 2: ❗ It's not just about plugin authors, I'd just say "require" and "implementing class".

Reviewer 2: I think the "Notes" column should be renamed and expanded with more usage requirements. Some more background on these tags, that could be used here:

https://nipafx.dev/javadoc-tags-apiNote-implSpec-implNote/

`@apiSpec`
This is the one we know and love; a description that applies equally to all valid implementations of the method, including preconditions, postconditions, etc.

`@apiNote`
Commentary, rationale, or examples pertaining to the API.

`@implSpec`
This is where we say what it means to be a valid default implementation (or an overrideable implementation in a class), such as "throws UOE." Similarly this is where we'd describe what the default for `putIfAbsent` does. It is from this box that the would-be-implementer gets enough information to make a sensible decision as to whether or not to override.

`@implNote`
Informative notes about the implementation, such as performance characteristics that are specific to the implementation in this class in this JDK in this version, and might change. These things are allowed to vary across platforms, vendors and versions.

Author: Updated in the Style Guide. Thank you.

| 3 | `@implNote`    | Adds a “Implementation Note” heading         | Describe the implementation provided by Gradle   |

When a block tag doesn't fit on a single line, continuation lines are indented four (or more) spaces from the position of the `@`.

## 1.2 The summary fragment

> Each Javadoc block begins with a brief summary fragment. 

Reviewer 2: ```suggestion
Each Javadoc block begins with a brief summary fragment - this is the first sentence up until the character `.` is encountered. 
```

 * This method always returns immediately, whether or not the image exists. 
 *
 * @param url an absolute URL giving the base location of the image
 * @param name the location of the image, relative to the url argument
 * @return the image at the specified URL
 * @see Image
 */
public Image getImage(URL url, String name) {
   try {
       return getImage(new URL(url, name));
   } catch (MalformedURLException e) {
       return null;
   }
}
```

Basic formatting rules:

- The first line contains the begin-comment delimiter ( `/**`).
- The first sentence is a summary.
- Notice the inline tag `{@link URL}`, which converts to an HTML hyperlink pointing to the documentation for the URL class.
- If you have more than one paragraph in the doc comment, separate the paragraphs with a `<p>` paragraph tag, as shown.
- Insert a blank comment line between the description and the list of tags, as shown.
- The first line that begins with an `@` character ends the description; you cannot continue the description following block tags.
- Block tags must be added in order.
- The last line contains the end-comment delimiter ( `*/`).

> So lines won't wrap, limit any doc-comment lines to 80 characters.

Reviewer 2: Can we increase this to 100 or 120?  80 is so low and monitors are bigger these days.

### 1.1.2 HTML

> Javadoc permits only a subset of HTML tags: `"a", "abbr", "acronym", "address", "area", "b", "bdo", "big", "blockquote", "br", "caption", "cite", "code", "colgroup", "dd", "del", "dfn", "div", "dl", "dt", "em", "fieldset", "font", "h1", "h2", "h3", "h4", "h5", "h6", "hr", "i", "img", "ins", "kbd", "li", "ol", "p", "pre", "q", "samp", "small", "span", "strong", "sub", "sup", "table", "tbody", "td", "tfoot", "th", "thead", "tr", "tt", "u", "ul", "var"`.

Reviewer 2: This will probably change with newer versions of the javadoc tool.  Can we link to the source of this list here?

> Attempt to limit your usage of HTML to: `"h1", "h2", "h3", "h4", "em", "li", "ol", "p", "ul"`.

Reviewer 2: ```suggestion
Attempt to limit your usage of HTML to: `"a", "h1", "h2", "h3", "h4", "em", "li", "ol", "p", "ul"`.
```

Also, I propose that if you're including 4 levels of headers in a single javadoc comment, you're doing it wrong.

Reviewer 2: Sigh...lol.

HTML tags for other block-level elements, such as `<ul>` or `<table>`, are not preceded with `<p>`.

### 1.1.4 Symbols

Entities for the less than symbol (<) and the greater than symbol (>) should be written as `&lt`; and `&gt;`. 
Similarly, the ampersand (&) should be written as `&amp;`:

```java
/**
 * This &amp; that.
 */
```

### 1.1.5 Links

The `{@link <class or method reference>}` tag is specifically used to link to the Javadoc of other classes and methods:

```java
/**
 * The url argument must specify an absolute {@link classname}. The name
 */
```

To add a reference to an external URL, use `<a href="URL#value">label</a>`.
This adds a link as defined by `URL#value`. The `URL#value` is a relative or absolute URL:

```java
/**
 * This is a link to <a href="http://www.google.com">Google</a>. The name
 */
```

### 1.1.6 Code keywords, names, and variables

Use `<code>...</code>` or `{@code ... }` style for keywords and names including:
- Java keywords
- package names
- class names
- method names
- interface names
- field names
- argument names
- code examples (see section 1.3 below)

For example:

```java
/**
 * Use the <code>Project</code> instance to configure the project.

>  * Use the {@code Project} instance to configure the project.

Reviewer 2: I don't think this example is what we want to emulate.  I think we should favor the briefer `{@code SOMECODE}` syntax most of the time, except when special characters make that difficult, or we're doing a multiline snippet.  I think we need an example of a multiline snippet here.  

Why wouldn't `{@code Project}` use `@link` instead?

Author: @code just formats the text and @link links to an existing class in the remainder of your javadoc. @link is always prefered but if there is nothing to "link" to or if you want to highlight something like a very generic class you can use @code.
I updated the styleguide accordingly.

| 4 | `@see`        | Adds a “See Also” heading with a link or text entry that points to a reference | `@see string`<br>`@see <a href=”URL#value”>label</a>`<br>`@see package.class#member label` |
| 5 | `@since`      | Adds a “Since” heading                                                         | Include the Gradle version if applicable                                                   |
| 6 | `@deprecated` | Adds a comment indicating that this API should no longer be used               | Make sure to have an alternative API linked                                                |

Custom Gradle tags:

| # | Tag         | Usage                                        | Notes                                                                                                                                                                                                               |
|---|-------------|----------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | 1 | `@apiSpec`  | Adds a “API Requirements” heading            | A description that applies equally to all valid implementations of the method, including preconditions, postconditions, etc.                                                                                        |

Reviewer 2: These spec/note tags are to appear below `@deprecated`, correct?  They're not Custom Gradle tags (though they are less frequently needed) and should probably just be part of a single list.

Author: These became standard in the newer JDK but I don't think they are standard in JDK 11. I will update accordingly either way.

| 2 | `@apiNote`  | Adds a “API Note” heading                    | A commentary, rationale, or example pertaining to the API.                                                                                                                                                          |
| 3 | `@implSpec` | Adds a “Implementation Requirements” heading | This is where the default implementation (or an overrideable implementation in a class) is specified.                                                                                                               |
| 4 | `@implNote` | Adds a “Implementation Note” heading         | This section contains informative notes about the implementation, such as advice to implementors, or performance characteristics that are specific to the implementation in this class of this version of the JDK.  |

> When a block tag doesn't fit on a single line, continuation lines are indented four (or more) spaces from the position of the `@`.

Reviewer 2: Important point that's easy to miss.  Could it be worked into an example?

## 1.2 The summary fragment

Each Javadoc block begins with a brief summary fragment - this is the first sentence up until the character `.` is encountered.

This fragment is very important: it is the only part of the text that appears in certain contexts such as class and method indexes:

```java
/**
 * Returns an Image object that can then be painted on the screen.
 */
```

## 1.3 Code blocks and snippets

> `<pre>` is the default HTML tag for preformatted text.

Reviewer 5: I think our convention for multi-line snippets is actually `<pre>{@code ...... }</pre>` and that's what it's also recommended: https://reflectoring.io/howto-format-code-snippets-in-javadoc/

Author: Yes, that is the standard convention, but I added syntax highlighting for code snippets in the Javadocs, which requires this "custom" implementation.

All code blocks and snippets should be wrapped in `<pre>...</pre>` at minimum.

Code blocks are formatted using [`highlight.js`](https://highlightjs.org/) using the `<code>` tag.
All code blocks and snippets should be wrapped in `<code class="language-*****"></code>` to be automatically highlighted.
Make sure the correct language is added to the class name: `language-kotlin`, `language-groovy`, or `language-java`.

A formatted and highlighted code snippet looks should look as follows:

```java
/**

>  * <pre class='groovy autoTested'><code class="language-kotlin">

Reviewer 2: This should note the mechanism for automatic testing.  It's not enough to just add the `autoTested` class, your project also needs a test class that extends `AbstractAutoTestedSamplesTest` to run them.  Hopefully we'll get rid of this requirement soon.

There are also other "tag suffixes" that do different things, as visible here: https://github.com/gradle/gradle/blob/master/testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractAutoTestedSamplesTest.groovy#L39-L48

 * project.ext.prop1 = "foo"
 * task doStuff {
 *     ext.prop2 = "bar"
 * }
 * subprojects { ext.${prop3} = false }
 * </code></pre>
 */
```

In order to label the coding language used in the snippet, you can use additional HTML, `<div class="code-block"><span class="label">Groovy</span>...</div>`:

```java
/**
 * <div class="code-block">
 * <span class="label">Groovy</span>
 * <pre class='groovy autoTested'><code class="language-groovy">
 * defaultTasks('some-task')    // Delegates to Project.defaultTasks()
 * reportsDir = file('reports') // Delegates to Project.file() and the Java Plugin
 * </code></pre>
 * </div>
 */
```

## 1.4 Where Javadoc is used

At the minimum, Javadoc is present for every public class, and every public or protected member of such a class, with a few exceptions such as overrides and self-explanatory members:

```java
public Image getImage(URL url, String name) {}
```

> ## 1.5 A note on IDEs

Reviewer 4: Maybe it's worth adding another sub-section on how to best configure the IDE to make it easier to follow the style guide.

Things like https://github.com/gradle/gradle/pull/28712/files#r1619850293 should go in there, so that everyone can be pointed at those settings. That is, if they cannot be shared via a config file checked into git.
